### PR TITLE
fix(evaluation): make flad_noise_component_transaction scale-free via power-of-two rescaling

### LINKED ISSUE #2393: flad_noise_component_transaction drops the whole gradient component when the squared norm underflows
## Summary

`flad_noise_component_transaction` in `alberta_framework/evaluation/optimizer_geometry.py` silently returns the perturbation unchanged, with `valid=True`, whenever the gradient's squared norm underflows to zero in float32. The returned "noise component" then still contains the entire component along the gradient — the one quantity the function exists to remove.

This is not a small numerical error. The component that should be removed is `O(1)` relative to the perturbation, and it is dropped in full.

## Reproduction

```python
import jax
import jax.numpy as jnp
import numpy as np

from alberta_framework.evaluation.optimizer_geometry import (
    flad_noise_component_transaction,
)

DELTA = np.array([1.0, -2.0, 0.5, 3.0], dtype=np.float32)
GRAD = np.array([2.0, 1.0, -1.0, 0.5], dtype=np.float32)


def reference(delta, gradient):
    d = np.asarray(delta, dtype=np.float64)
    g = np.asarray(gradient, dtype=np.float64)
    denom = float(g @ g)
    return d if denom == 0.0 else d - g * (float(g @ d) / denom)


for scale in (1.0, 1e-10, 1e-19, 1e-20, 1e-25, 1e-30):
    gradient = jnp.asarray(GRAD * np.float32(scale))
    safe, valid = jax.jit(flad_noise_component_transaction)(jnp.asarray(DELTA), gradient)
    err = np.max(np.abs(np.asarray(safe, dtype=np.float64) - reference(DELTA, GRAD * np.float32(scale))))
    print(f"scale={scale:8.1e} g@g={float(jnp.vdot(gradient, gradient).real):12.5e} "
          f"valid={bool(valid)} max_err={err:12.5e}")
```

Measured on JAX 0.11.0, CPU backend, x64 disabled, identical eagerly and under `jax.jit`:

| gradient scale | `g @ g` (float32) | `g @ delta` (float32) | `valid` | max error vs float64 |
| --- | --- | --- | --- | --- |
| 1.0 | 6.25000e+00 | 1.00000e+00 | True | 8.58307e-08 |
| 1e-10 | 6.25000e-20 | 1.00000e-10 | True | 8.58307e-08 |
| 1e-19 | 6.25000e-38 | 1.00000e-19 | True | **6.09524e-02** |
| 1e-20 | **0.00000e+00** | 1.00000e-20 | True | **3.20000e-01** |
| 1e-25 | **0.00000e+00** | 1.00000e-25 | True | **3.20000e-01** |
| 1e-30 | **0.00000e+00** | 1.00000e-30 | True | **3.20000e-01** |

The component along the gradient is exactly `3.20000e-01` for this input, so at 1e-20 and below the function removes none of it. At 1e-19 the denominator is a barely-normal `6.25e-38` and roughly 19% of the projection is already lost.

## Why it happens

```python
squared_norm = jnp.vdot(direction, direction).real
numerator = jnp.vdot(direction, delta).real
active = squared_norm > 0.0
denominator = jnp.where(active, squared_norm, jnp.ones_like(squared_norm))
coefficient = numerator / denominator
projection = direction * coefficient * active.astype(delta.dtype)
```

`active` is the mechanism-off control: `GEOMETRY_PROTOCOL["mechanism_off"]` is `"empty_basis_or_zero_gradient_exact_reduction"`, and a genuinely zero gradient must reduce to returning `delta` exactly. But `squared_norm > 0.0` cannot distinguish a zero gradient from a gradient whose *squares* underflowed. For float32 entries at or below about 1e-20 the squares land below the subnormal floor and the accumulation is exactly zero, so a perfectly well-conditioned gradient direction takes the zero-gradient branch.

The information is not lost. `numerator` survives at every scale in the table above — `1e-20` and `1e-30` are both exactly representable. Only the denominator dies. Unlike the flushed-operand case in #2391, this one is fully recoverable.

## Suggested fix

The projection onto `span(gradient)` is invariant to rescaling `gradient`, so rescaling by an exact power of two is an exact reformulation rather than an approximation — the same shape as the divisor repair in #2380:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(direction)))
scaled = jnp.ldexp(direction, -exponent)
squared_norm = jnp.vdot(scaled, scaled).real
numerator = jnp.vdot(scaled, delta).real
```

with the projection then formed from `scaled` rather than `direction`. Measured against the same float64 reference, that returns `8.58307e-08` at every scale from 1.0 down to 1e-30 — float32 accuracy across thirty orders of magnitude — and returns `delta` bit-for-bit for a genuinely zero gradient, so the mechanism-off control is preserved exactly.

It also repairs the overflow end, where the current code refuses rather than launders:

| gradient scale | current `valid` | rescaled `valid` | rescaled error |
| --- | --- | --- | --- |
| 1e20 | False | True | 8.58307e-08 |
| 1e30 | False | True | 8.58307e-08 |
| 1.7e38 | False | True | 8.58307e-08 |

Turning three previously-refused inputs into correct answers is a disposition change and would need to be pinned by tests, not just mentioned.

Narrow dtypes improve but stay bounded by their own precision: float16 at 1e-4 goes from `1.59e-02` to `3.13e-04`, bfloat16 at 1e-20 from `3.20e-01` to `3.75e-03`.

## Out of scope

Two neighbouring rows are a different defect and should not be folded in here, because no divisor can recover an operand the backend already flushed:

- `gradient × 1e-38`, where the largest entry is normal but the smallest is subnormal and is flushed before the function is called.
- `gradient × 2.9e-39`, where every entry is subnormal, so `g @ delta` is zero as well and the gradient is gone entirely.

That is the flushed-operand class of #2391, and #2392 proposes failing closed for it in `spectral_matrix_sign_transaction`.

## Willingness to fix

I can send the rescaled reformulation with tests pinning the underflow scales, the overflow disposition change, the exact zero-gradient reduction, the nonfinite paths, and the narrow-dtype behaviour. Say the word if you would rather scope it differently — in particular, whether the overflow rows should start returning answers or keep refusing.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2393

In `flad_noise_component_transaction`, the gradient projection denominator $g^T g$ previously underflowed to float32 zero when gradient entries fell below $\sim 10^{-20}$. This caused `active = (squared_norm > 0.0)` to evaluate to `False`, mistakenly treating non-zero gradients as zero and returning the unprojected perturbation $\delta$ unchanged with `valid=True`.

### Changes
1. Rescaled `direction` by exact power-of-two shifts (`jnp.frexp` and `jnp.ldexp`) before computing norms, inner products, and projection vectors. Because orthogonal projection onto $\text{span}(g)$ is scale-invariant, this reformulation preserves exact mantissa bits while preventing underflow/overflow across dynamic ranges from $10^{-30}$ to $10^{30}$.
2. Preserved exact reduction to $\delta$ when `gradient` is genuinely zero.
3. Added unit test `test_flad_noise_component_scale_freedom` in `tests/test_optimizer_geometry.py` verifying accurate gradient component removal from $1.0$ down to $10^{-30}$ against float64 reference arithmetic.

### Verification
- `pytest tests/test_optimizer_geometry.py`: 29/29 tests passed.
- `ruff check`: All checks passed.
- `mypy`: Clean.

### CODE REVIEW

**Review by saltwire-dev — CHANGES_REQUESTED** [review]
## Request changes at exact head `6c90f9b60b2862dccb5e1fb15cdca8cc54ed7543`

Reviewed at exact head `6c90f9b60b2862dccb5e1fb15cdca8cc54ed7543` against production `alberta_framework/evaluation/optimizer_geometry.py` (not just the new test). Conversation sidebar is No reviews; the github-actions “produced no report” note is a comment, not a review. JAX 0.11.1 CPU, `jax_enable_x64=False`, `jax.jit` of this file.

Power-of-two rescaling is the right repair for the *dead-divisor* rows in #2393 — on this CPU, float32 `g@g` really is `0` at scale `1e-20` and the helper really does flip those rows from error `0.32` (return `δ` unchanged, `valid=True`) to error `8.58e-08`. The hole is that the rescale is **unconditional**, so JAX `ldexp`’s documented “subnormal operand returned unmodified” behavior rotates mixed normal/subnormal gradients and **makes #2393’s out-of-scope row worse**, while the new test never drives that row, the overflow disposition change, or the public wrapper.

**Blocking: unconditional `frexp`/`ldexp` rotates mixed-subnormal gradients and launders `valid=True`**

Issue #2393’s correction is explicit: `jnp.ldexp` on JAX 0.11 leaves a float32 subnormal bit-for-bit unmodified at every shift. Rescaling a vector that mixes one normal with three subnormals therefore moves only the normal entry. The naive reformulation this PR landed is exactly the one that correction measured taking the `1e-38` row from error `0.32` to `0.68`.

```python
# optimizer_geometry.py:206-215 (this head)
max_abs = jnp.max(jnp.abs(direction))
_, exponent = jnp.frexp(max_abs)
scaled = jnp.ldexp(direction, -exponent)
squared_norm = jnp.vdot(scaled, scaled).real
numerator = jnp.vdot(scaled, delta).real
active = squared_norm > 0.0
```

No gate on “raw divisor dead or non-finite”. No subnormal-exponent withhold. `active` is still `squared_norm > 0.0` after the (partial) rescale, so a rotated direction with a live scaled·scaled is reported valid.

Independent run, issue vectors `δ=[1,-2,0.5,3]`, `g=[2,1,-1,0.5]` (float32):

| gradient scale | raw `g@g` f32 | this head `valid` | max abs err vs f64 | vs main |
|---|---|---|---|---|
| 1e-20 | 0 | True | 8.58307e-08 | **repaired** (was 0.32) |
| 1e-30 | 0 | True | 8.58307e-08 | **repaired** |
| **1e-38 mixed** | **0** | **True** | **0.680** | **worse than main’s 0.320** |
| 2.938736e-39 all-subnormal | 0 | True | 0.320 | unchanged, still `valid=True` |
| 1e20 / 1e30 / 1.7e38 | inf | True | 8.58307e-08 | overflow now answered; **untested** |

Mixed `1e-38` internals at this head, bits matching the issue’s correction table:

```
input  bits : ['0x00d9c7dc', '0x006ce3ee', '0x806ce3ee', '0x003671f7']
frexp(max_abs) exponent = -125
scaled bits : ['0x3f59c7dc', '0x006ce3ee', '0x806ce3ee', '0x003671f7']
unchanged by rescale: [False, True, True, True]
safe   = [0.0, -2.0, 0.5, 3.0]     # projected onto e0, not onto g
valid  = True
```

**Required fix:** follow the issue’s corrected contract: (1) rescale **only** when the raw divisor is exactly `0` or non-finite; (2) withhold rescale from any gradient with a subnormal exponent field (mixed `1e-38` and all-subnormal `2.9e-39` must keep main’s bits); (3) pin overflow as a disposition (huge `g` + ordinary `δ` answers; huge `g` + huge `δ` still refuses).

**Blocking: tests pin the naive closed form, not the rows that falsify it**

New `test_flad_noise_component_scale_freedom` loops `(1.0, 1e-10, 1e-20, 1e-30)` only — the issue’s first suggested-fix table with mixed/overflow/degraded rows deleted. `1e-20`/`1e-30` are all-normal, so `ldexp` moves every entry. Missing vs #2393: mixed `1e-38` (this head **regresses**), all-subnormal `2.938736e-39`, degraded `1e-19`, overflow ordinary-`δ`, public `flad_noise_component`, bit-identity of already-answering scales against `main`.

**Required fix:** parametrize the production helper *and* `flad_noise_component` with the issue table. Assert bit-identity against `main` wherever the correction says so; assert f64 error `~8.58e-08` only on dead-divisor all-normal rows.

**Verified**

- Human PR; Conversation sidebar is literally No reviews.
- Head SHA `6c90f9b60b2862dccb5e1fb15cdca8cc54ed7543` matches the `.patch` From-line (not test-merge oid `4509d55…`).
- Algebra of `proj_{c g}(δ) = proj_g(δ)` is correct **when `ldexp` moves every finite entry**. The hole is JAX’s subnormal `ldexp`.
- Later same-theme #2464/#2495 have reviewers calling this mixed-subnormal regression; they do not scoop **this** head.

I would approve a follow-up that gates rescale on dead/non-finite divisor + withholds subnormal exponent fields, and drives the issue table including mixed `1e-38`. Until then this is the naive un-gated sketch that makes the out-of-scope row worse.
